### PR TITLE
feat: Introduce cache controller to support GC for workflow memoization caches

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes https://github.com/argoproj/argo-workflows/issues/5406. Opening this in draft for earlier feedback and design discussions.

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

**Comment by terrytangyuan** [comment]
I opened this draft PR so we can discuss whether I am in the right direction. No code-gen or tests yet so it's easier review. 

@simster7 @alexec @sarabala1979 and others who are interested. Any suggestions/feedback would be appreciated.

**Comment by terrytangyuan** [comment]
@simster7 May I ask for your review on the direction of this? We have something similar to this running internally for a while now and it's been working very well. It would be great to get your feedback before I spend time cleaning up the code and adding tests, etc.

**Comment by alexec** [comment]
Still in draft

File: workflow/controller/cache/controller.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+func NewController(namespace string, kubeclientset kubernetes.Interface, metrics *metrics.Metrics, eventRecorderManager events.EventRecorderManager) *Controller {
+	return &Controller{
+		namespace:            namespace,
+		kubeclientset:        kubeclientset,
+		cacheQueue:           metrics.RateLimiterWithBusyWorkers(workqueue.DefaultControllerRateLimiter(), "cache_queue"),
+		eventRecorderManager: eventRecorderManager,
+	}
+}
+
+func (cc *Controller) Run(ctx context.Context) {
```
Comment (alexec) [inline workflow/controller/cache/controller.go:48]: I think you could simplify this code with something like:

```
v1.NewFilteredPodInformer(k, managedNamespace, podResyncPeriod, indexers, func(opts *metav1.ListOptions) {
			opts.LabelSelector = labelSelector
		})
```

File: workflow/controller/cache/controller.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+			DeleteFunc: func(obj interface{}) {
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err == nil {
+					cc.cacheQueue.Add(key)
+				}
+			},
+		})
+	log.Info("Watching config map updates")
+	go cc.cacheController.Run(ctx.Done())
+	go wait.UntilWithContext(ctx, cc.syncAll, 10*time.Second)
```
Comment (alexec) [inline workflow/controller/cache/controller.go:98]: resync is automatic with an informer

**Comment by terrytangyuan** [comment]
Closing this for now and will be done in separate PRs after our discussion.

**Comment by agilgur5** [comment]
Seems like this was superseded by #6850

**Pull request closed without merging**